docs/source/notes/source_of_recompilation.md

```diff
@@ -0,0 +1,176 @@
+* parse python source. code to get the control flow statement automatically. This is like Torchscript and somehow merge the torchscripted graph into the lazily trace graph properly (including shape info etc). I haven’t thought through the steps of how to implement this indeed :P
+
+But either solution above requires non-trivial amount of effort, either on user side or on the framework side. That’s why we currently just take the hit of early evaluation & multiple compilations as a short term solution given the bandwidth we have.
+
+Okay so now we assume that also have control flow lowered in the graph automagically, are we gold?
+
+YES! Now you have your whole computation represented in a graph of Tensor operations, including control flow so that compilers can now consume and do their smart tricks! But tbh at this point your program is no longer very PyTorch-y.
+
+
+## Conclusion:
+
+There’re actually multiple sources of recompilation and bounded dynamic shape support cannot solve all of them. The proposed workarounds in this doc are definitely sometimes impractical, and there might be better ways to fix each source properly that I’m totally unaware of. But I hope as we keep smashing our way to an ideal lazy tensor stack in this doc, it’s now easier for you understand what’re the remaining blockers ahead of us.
+
+
+## Appendix:
+
+1. NNC uses symbolic shapes, does that help?
+
+Yes but partially. By having symbolic shape, your compilation optimization no longer requires concrete shape values. In other words your generated kernel are more general than XLA’s static shape ones.
+
+And which exactly problem does it help?
+
+It helps with cases like #1 and #2.1.
+
+
+```
+shape [3, 5] -> add -> transpose -> ... -> mul
+shape [6, 2] -> add -> transpose -> ... -> mul
+
+# with symbolic shape
+shape [x, y] -> add -> transpose -> ... -> mul
+```
+
+With symbolic shape your generated kernel doesn’t recompile as XLA does with static shapes.
+
+XLA solves this problem in the other way, by using padding/bucketization (for #1) and bounded dynamic shape (for #2.1).
+
+Brian Hirsh(@bdhirsh) asked some really good questions in the comment, moving here to make them more visible:
+
+2. Is it worth sticking a TORCH_WARN in the XLA kernels of ops that produce data-dependent output shapes?
+
+Yea torch_warn is useful in telling users "hey your program won't run blazing fast". But for these data dependent ops, there isn't an easy rewrite for them unless users change the logic in their model. (another example is torch.unique())
+
+3. How ops like nonzero impact our ability to devirtualize sizes()? If we want to devirtualize sizes(), we’ll need to be able to eagerly compute sizes for each op - won’t that mean we’re forced to evaluate the graph every time we hit an op like nonzero? Vs. right now, it sounds like we don't actually force an evaluation when a user calls nonzero()?
```

Comment:
@ailzhang Do you have a little bit more context for the intent to devirtualize sizes()? 

Leave torch/xla aside, current cuda implementation of nonzero actually triggers a host-device sync, and that cost would be more substantial than a devirtualization of a virtual function call.